### PR TITLE
Add `ispositive`, etc.

### PR DESCRIPTION
**Pull request by putianyi889**
- https://github.com/JuliaLang/julia/issues/35067

It's faster for some complicated Real types.
```julia
julia> f(x) = x<=zero(typeof(x))
f (generic function with 1 method)

julia> g(x) = signbit(x)||iszero(x)
g (generic function with 1 method)

julia> @btime f(x) setup=(x=rand(BigFloat))
  99.055 ns (4 allocations: 144 bytes)
true

julia> @btime g(x) setup=(x=rand(BigFloat))
  10.700 ns (0 allocations: 0 bytes)
false
```

If you think this is a good idea then I'll add docstrings and tests.

Edit: FAQs according to discussions below
- Why `ispositive(x::Bool) = x`?
  - Because it's fast. https://github.com/JuliaLang/julia/pull/53677#discussion_r1530731988
- Why support `ispositive(::Any)`?
  - #35067

TODO:
- [x] Edit `NEWS.md`.
- [x] Add a comment on future change as #21712 is closed. ref: https://github.com/JuliaLang/julia/pull/53677#discussion_r1534394959

### CODE REVIEW

**Comment by goerz** [comment]
For additional context: https://github.com/JuliaRegistries/General/pull/102561

File: base/number.jl [hunk 1]
```diff
@@ -136,6 +136,11 @@ true
 """
 signbit(x::Real) = x < 0
 
+ispositive(x) = !isnonnegative(x)
+isnegative(x) = signbit(x) && !iszero(x)
```
Comment (giordano) [inline base/number.jl:140]: Why not 
```suggestion
isnegative(x) = !isnonpositive(x)
```
as you did above?

Comment (putianyi889) [reply base/number.jl:140]: because there is one more operation? tbh I don't know if Julia does the optimization.

Comment (giordano) [reply base/number.jl:140]: > tbh I don't know if Julia does the optimization.

I'm away from computer, but I'm pretty sure it'd optimise away double negations. And can you please also comment on the second on the second question. I don't really see the need for 4 new functions when 2 would be enough.

Comment (putianyi889) [reply base/number.jl:140]: > it'd optimise away double negations

If that's the case then 2 is enough.

Comment (barucden) [reply base/number.jl:140]: Wait, "nonpositive" does not mean "negative"; it means "negative or zero" (and similarly for "nonnegative"). So, there still might be some value in providing the definitions
```julia
isnonpositive(x) = isnegative(x) || iszero(x)
isnonnegative(x) = ispositive(x) || iszero(x)
```

**Comment by mikmoore** [comment]
I'll just remark on `ispositive`, but the same goes for `isnegative`.

Needs NaN handling. For example,
```julia-repl
julia> ispositive(NaN) ⊻ ispositive(-NaN)
true
```
A possible fix is in the flavor of `ispositive(x) = !isnan(x) && !signbit(x) && !iszero(x)`. The default for non-float `Number` types is `isnan(x) = false` so this should be compiled away when irrelevant.

Also needs proper `Missing` behavior, `ispositive(::Missing) = missing`.

Now for your benchmarks (which appear to implement a "isnonpositive" function that does not appear in this PR, but I'll roll with it for now):
```julia-repl
julia> using BenchmarkTools

julia> f(x) = x<=zero(typeof(x));

julia> g(x) = signbit(x)||iszero(x);

julia> g_fixed(x) = !isnan(x)&&(signbit(x)||iszero(x));

julia> h(x) = x <= false;

julia> h_int(x) = x <= 0;

julia> vi32 = rand(Int32,1000); vf64 = randn(Float64,1000); vbf = randn(BigFloat,1000);

julia> @btime count(f,$vi32); @btime count(g,$vi32); @btime count(g_fixed,$vi32); @btime count(h,$vi32); @btime count(h_int,$vi32)
  84.528 ns (0 allocations: 0 bytes)
  84.112 ns (0 allocations: 0 bytes)
  84.495 ns (0 allocations: 0 bytes)
  84.079 ns (0 allocations: 0 bytes)
  82.622 ns (0 allocations: 0 bytes)
517

julia> @btime count(f,$vf64); @btime count(g,$vf64); @btime count(g_fixed,$vf64); @btime count(h,$vf64); @btime count(h_int,$vf64)
  61.914 ns (0 allocations: 0 bytes)
  90.900 ns (0 allocations: 0 bytes)
  134.514 ns (0 allocations: 0 bytes)
  59.470 ns (0 allocations: 0 bytes)
  58.163 ns (0 allocations: 0 bytes)
489

julia> @btime count(f,$vbf);  @btime count(g,$vbf);  @btime count(g_fixed,$vbf);  @btime count(h,$vbf);  @btime count(h_int,$vbf)
  25.100 μs (2000 allocations: 101.56 KiB)
  12.900 μs (0 allocations: 0 bytes)
  14.000 μs (0 allocations: 0 bytes)
  54.900 μs (2000 allocations: 39.06 KiB)
  14.200 μs (0 allocations: 0 bytes)
498
```
Don't get hung up on differences of a few ns, which are probably benchmarking artifacts. But notice that this function is notably slower for native floats (and *much* slower when correctly handling NaN). And it doesn't out-do simple comparisons for `BigFloat` (that avoid creating a new `BigFloat`).

My preferred implementation would be `h`, although we need to add proper inequality handling for `_comparison_(BigFloat, Bool)` to get it to match the `h_int` performance in that case.

**Comment by putianyi889** [comment]
> I'll just remark on `ispositive`, but the same goes for `isnegative`.

Thanks for the investigation! Looks like the problem is more complex than what I've imagined. Are you on the dev version of Julia? My benchmark results are a little different. I'm on v1.10.0.

File: test/numbers.jl [hunk 2]
```diff
@@ -830,6 +830,47 @@ end
...
 end
+@testset "ispositive/isnegative" begin
+    @test ispositive(1)        == true
+    @test ispositive(0)        == false
+    @test ispositive(-1)       == false
+    @test ispositive(1.0)      == true
+    @test ispositive(0.0)      == false
+    @test ispositive(-0.0)     == false
+    @test ispositive(-1.0)     == false
+    @test ispositive(1.0/0.0)  == true
```
Comment (giordano) [inline test/numbers.jl:841]: This is redundant with `ispositive(Inf)` below

File: test/numbers.jl [hunk 3]
```diff
@@ -830,6 +830,47 @@ end
     @test cmp(isless, 1, NaN) == -1
     @test cmp(isless, NaN, NaN) == 0
 end
+@testset "ispositive/isnegative" begin
```
Comment (giordano) [inline test/numbers.jl:833]: It'd be better to test in a for loop multiple types (e.g. all `IEEEFloat`, and similarly for the standard integers), instead of testing just two types.

File: test/numbers.jl [hunk 4]
```diff
@@ -830,6 +830,47 @@ end
...
+    @test ispositive(1)        == true
+    @test ispositive(0)        == false
+    @test ispositive(-1)       == false
+    @test ispositive(1.0)      == true
+    @test ispositive(0.0)      == false
+    @test ispositive(-0.0)     == false
+    @test ispositive(-1.0)     == false
+    @test ispositive(1.0/0.0)  == true
+    @test ispositive(-1.0/0.0) == false
```
Comment (giordano) [inline test/numbers.jl:842]: This is redundant with `ispositive(-Inf)`

File: test/numbers.jl [hunk 5]
```diff
@@ -830,6 +830,47 @@ end
     @test cmp(isless, 1, NaN) == -1
     @test cmp(isless, NaN, NaN) == 0
 end
+@testset "ispositive/isnegative" begin
+    @test ispositive(1)        == true
```
[the remark this answers is not in the document]
Comment (putianyi889) [reply test/numbers.jl:834]: I referred to 
https://github.com/putianyi889/julia/blob/b4266e897d0a74d828906b4b2b7f4999786a8b66/test/numbers.jl#L684-L717

If you prefer your style then I could change that as well.

File: test/numbers.jl (hunk 2 above)
Comment (putianyi889) [reply test/numbers.jl:841]: I copied the tests from
https://github.com/putianyi889/julia/blob/b4266e897d0a74d828906b4b2b7f4999786a8b66/test/numbers.jl#L538-L556
same holds for some of your comments below.

Do you think it's a good idea to use a different style than the rest of the file?

Comment (giordano) [reply test/numbers.jl:841]: Well, yes, I don't see how `f(1.0/0.0)` is testing anything different than `f(Inf)`.

Comment (putianyi889) [reply test/numbers.jl:841]: Would you like to delete those from other tests?

**Comment by simonbyrne** [comment]
One thought: rather than defining these, would it be better to define a function which returns a Real or Number, say `value`?

**Comment by barucden** [comment]
Don't forget to add a compatibility note for the added functions (they are likely to be a part of Julia 1.12.0).

File: test/numbers.jl [hunk 6]
```diff
@@ -830,6 +830,54 @@ end
     @test cmp(isless, 1, NaN) == -1
     @test cmp(isless, NaN, NaN) == 0
 end
+@testset "ispositive/isnegative" begin
+    for T in Base.uniontypes(Base.IEEEFloat)
```
Comment (giordano) [inline test/numbers.jl:834]:
```suggestion
@testset "ispositive/isnegative" begin
    @testset "Type $(T)" for T in Base.uniontypes(Base.IEEEFloat)
```
and same below, so that people can more easly tell what type is failing

**Comment by mikmoore** [comment]
Would it make more sense for `x > 0` (etc) to be the default implementation and the `!isnan & !iszero & !signbit` to be the specialization for `BigFloat` and maybe `::BigInt` (basically, to swap the `IEEEFloat` specialization and the `::Any` fallback?) As far as I can tell, `BigFloat` (and maybe ::BigInt) is the only `Base` type that will be handled better by this function.

I still think that it might be worth exploring faster inequality checks for `BigFloat` mixed with other types (at least `Bool` should be easy). It's a little more complicated to implement, but it could make these new functions unnecessary. I don't see these new functions being used especially often. They add a faster alternative to a specific `BigFloat` operation, but otherwise are equivalent to `x -> x>0` etc. `BigFloat` is already very slow so I don't see this making a big performance difference in real code.

File: base/number.jl [hunk 7]
```diff
@@ -136,6 +136,45 @@ true
...
+
+julia> isnegative(99)
+false
+
+julia> isnegative(-0.0)
+false
+```
+"""
+isnegative(x) = !isnan(x) && signbit(x) && !iszero(x)
+isnegative(x::Integer) = signbit(x) # skip iszero
```
Comment (mikmoore) [inline base/number.jl:176]: Nothing about `Integer` guarantees that `-0` isn't a valid number. That property is specific to `Base.BitInteger` and `BigInt`. I don't think this specialization is valid.

File: base/floatfuncs.jl [hunk 8]
```diff
@@ -16,6 +16,9 @@ signbit(x::Float64) = signbit(bitcast(Int64, x))
 signbit(x::Float32) = signbit(bitcast(Int32, x))
 signbit(x::Float16) = signbit(bitcast(Int16, x))
 
+ispositive(x::IEEEFloat) = x>0
+isnegative(x::IEEEFloat) = x<0
```
Comment (mikmoore) [inline base/floatfuncs.jl:20]: I would suggest we make these the default implementations. They are much more intuitive. If the point of this PR is to make `BigFloat` faster, focus on that.

File: base/number.jl (hunk 7 above)
Comment (putianyi889) [reply base/number.jl:176]: You could also argue that integer can be `Inf` and `NaN` as well (e.g. Infinities.jl), but they are not handled by the Julia base. Do you have an example of integer `-0`?

Comment (mikmoore) [reply base/number.jl:176]: No. But I could go create `SignedMagnitudeIntegers.jl` today and have a type with `-0`. See [Signed magnitude representation](https://en.wikipedia.org/wiki/Signed_number_representations#Sign%E2%80%93magnitude).

File: base/floatfuncs.jl (hunk 8 above)
Comment (putianyi889) [reply base/floatfuncs.jl:20]: The point is that `ispositive` and `isnegative` can work without comparison operators. For a custom type without hardware support, it's usually easier and faster to implement `signbit` and `iszero`.

`BigInt` and `isnegative(::Unsigned)` should also be faster.

File: test/numbers.jl (hunk 6 above)
Comment (putianyi889) [reply test/numbers.jl]: reopened since other tests need this

File: base/number.jl (hunk 7 above)
Comment (putianyi889) [reply base/number.jl:176]: The fallback for `signbit(::Integer)` is https://github.com/JuliaLang/julia/blob/a9611ce6acde605c4f5ec5787514e181b6b4add6/base/int.jl#L139

so I'd argue that my implementation is consistent. The drawback is that if you have a type like that, then you need to specialize `isnegative` and `signbit` at the same time, so basically one more task (compared to only `signbit`).
The advantage is that for integer types that don't have `-0`, one can either have a shortcut for `signbit` which makes `isnegative` faster, or it will work identically as `isnegative(x) = x<0`.

**Comment by putianyi889** [comment]
> As far as I can tell, `BigFloat` (and maybe ::BigInt) is the only `Base` type that will be handled better by this function.

I can think of `AbstractIrrational` (although `signbit` has not yet been specialized) and `Unsigned` as well, so basically `IEEEFloat` is the only group where `isnegative(x)=x<0` is strictly better (and `SignedBitInteger` is on par) and it's because of hardware support.

File: base/floatfuncs.jl (hunk 8 above)
Comment (mikmoore) [reply base/floatfuncs.jl:20]: These functions only make sense for `Real` anyway (and one could argue that they maybe be restricted as such? I'd suggest this except for the `Dual` type below.). Is it a realistic case that you implement a `Real` type where you don't care about comparisons but *do* care about signs? Both seems pretty fundamental to real numbers.

Is `signbit` really a fundamental function? More-so than comparison? Note that the most general definition is `signbit(x::Real) = x < 0`, so it uses comparison unless given an explicitly better option. In the meantime,
```julia-repl
julia> using DualNumbers, BenchmarkTools

julia> ltzero(x) = x<0
ltzero (generic function with 1 method)

julia> isnegative(x) = !isnan(x) && signbit(x) && !iszero(x)
isnegative (generic function with 1 method)

julia> d = Dual.(randn(1000));

julia> @btime count(ltzero, $d)
  179.714 ns (0 allocations: 0 bytes)
491

julia> @btime count(isnegative, $d)
ERROR: MethodError: no method matching signbit(::Dual128)
```
Now, maybe `Dual` should support `signbit` (for wrapped `Real` types, although generally `Dual <: Number` so it doesn't make sense). But here's a living example from the ecosystem that would not work with a `signbit` implementation.

`Unsigned` is an abstract type, but for any `BitUnsigned` you'll get
```julia-repl
julia> code_llvm(x -> x < 0, (UInt,))
;  @ REPL[35]:1 within `#27`
define i8 @"julia_#27_348"(i64 zeroext %0) #0 {
top:
  ret i8 0
```
so this is no help. If you want to make `BigInt` faster, go ahead and specialize them too.

**Comment by mikmoore** [comment]
> `Unsigned` as well

`Unsigned` is an abstract type, so you can't say anything about what the fastest implementation is. But for any `BitUnsigned` you'll get
```julia-repl
julia> code_llvm(x -> x < 0, (UInt,))
;  @ REPL[35]:1 within `#27`
define i8 @"julia_#27_348"(i64 zeroext %0) #0 {
top:
  ret i8 0
}
```
(aka `return false`) so this is no help. If you want to make `BigInt` faster, go ahead and specialize them too. The point of the fallback is to be general, not fast. Speed is the domain of type-specific specializations.

`IEEEFloat`, the direct comparison is fastest and `BitInteger` is tied between the two (because `isnegative` falls back to comparison).

You've shown that this function is faster for `BigInt` and `BigFloat`. I concede this. Although I fail to imagine where this makes a notable performance difference as any computation with these types is profoundly slower than these new comparisons. These types are slow because almost every operation with them makes foreign calls to an external library. Even if we make this comparison essentially free, one has presumably one actual calculation associated with each value.

Here's a microbenchmark (to expand on original benchmark that showed these new comparisons are much faster) that shows that virtually any real-code use of this will not significantly improve the speed.
```julia-repl
julia> using BenchmarkTools

julia> xb = big.(rand(Int,1000));

julia> @btime count(z->isnegative(z+1), $xb)
  63.400 μs (3000 allocations: 46.88 KiB)
506

julia> @btime count(z->(z+1)<0, $xb)
  66.800 μs (3000 allocations: 46.88 KiB)
506
```

File: base/floatfuncs.jl (hunk 8 above)
Comment (putianyi889) [reply base/floatfuncs.jl:20]: > Is it a realistic case that you implement a `Real` type where you don't care about comparisons but _do_ care about signs?

Some types are not easy/able to compare but we can easily check that they are positive or negative. For example, irrationals (that's my motivation) and dates (from https://github.com/JuliaLang/julia/issues/35067).

On the second case, it's suggested to use `x>zero(x)` and I think this should be the general fallback. For `Real` it should just be `x>0`.

> Is signbit really a fundamental function?

It should be on the same level as `iszero`. Its name might be bad, but the fallback suggests it's basically an `almost_isnegative` and its documentation says
`Return true if the value of the sign of x is negative, otherwise false.`

Comment (mikmoore) [reply base/floatfuncs.jl:20]: `signbit` is identical to `<(0)` except for its treatment of "negative zero" and NaN. As such, I consider it to be a function for bit-twiddling or explicitly checking the sign *bit* (where one wants to group +0 and +NaN with the positives and -0 and -NaN with the negatives, except that one usually doesn't actually care about the signs of NaNs). In other words, I consider `signbit` to be a statement about implemented representation rather than value.

I consider `sign` or `<(0)` to be "mathematical" functions that concern values rather than representation. Although `sign` still distinguishes +0 and -0, both are usually `==(0)`. I acknowledge that a version of `isnegative` for `BigInt` and maybe some other types are currently more efficient with `signbit`, but that's because of internal representation (which is why I advocate a specialization for such cases). Note that if we did (I doubt we would on any <5yr timescale) replace the `BigInt`/`BigFloat` libraries with native Julia implementations that handle mixed-type comparisons, this performance difference might reverse.

So personally, I consider `signbit` (a representation-testing function that happens to align with a value-testing function for some types) to be subordinate to meaning-based tests like `iszero` or `sign` or `<`. When writing operational code, I contend that one is more interested in meaning than representation. This is why I prefer `x -> x<0` as the generic implementation rather than the `signbit` implementation.

Note: although I often write `<(0)`, one would usually implement this as `x -> x<0` because `<(0)` is a `Base.Fix2` type and that often blocks propagation of the constant.

**Comment by putianyi889** [comment]
Just found

https://github.com/JuliaLang/julia/blob/master/base/gmp.jl#L735-L736

Should we continue to use `isnegative` and `ispositive` or adapt `isneg` and `ispos`?

File: base/number.jl [hunk 9]
```diff
@@ -136,6 +136,44 @@ true
...
+julia> ispositive(0.0)
+false
+```
+"""
+ispositive(x) = x>0
+
+"""
+    isnegative(x)
+
+Test whether `x` is positive. Falls back to [`isnan`](@ref), [`signbit`](@ref) and [`iszero`](@ref). See also [`ispositive`](@ref).
```
Comment (giordano) [inline base/number.jl:161]:
```suggestion
Test whether `x` is negative. Falls back to [`isnan`](@ref), [`signbit`](@ref) and [`iszero`](@ref). See also [`ispositive`](@ref).
```

File: base/gmp.jl [hunk 10]
```diff
@@ -912,7 +912,7 @@ module MPQ
 
 # Rational{BigInt}
 import .Base: unsafe_rational, __throw_rational_argerror_zero
-import ..GMP: BigInt, MPZ, Limb, isneg, libgmp
+import ..GMP: BigInt, MPZ, Limb, isnegative, libgmp
```
Comment (giordano) [inline base/gmp.jl:915]: I don't think you need to import `isnegative` now that's in `Base`?

Comment (putianyi889) [reply base/gmp.jl:915]: I'll need to check all presence of `isneg` and `ispos` first.

File: base/mpfr.jl [hunk 11]
```diff
@@ -1062,6 +1062,10 @@ isfinite(x::BigFloat) = !isinf(x) && !isnan(x)
 iszero(x::BigFloat) = x.exp == mpfr_special_exponent_zero
 isone(x::BigFloat) = x == Clong(1)
 
+# In theory, `!iszero(x) && !isnan(x)` should be the same as `x.exp > mpfr_special_exponent_nan`, but this is safer.
+ispositive(x::BigFloat) = !signbit(x) && !iszero(x) && !isnan(x)
+isnegative(x::BigFloat) = signbit(x) && !iszero(x) && !isnan(x)
```
Comment (giordano) [inline base/mpfr.jl:1067]: I presume you need to import these functions from `Base` to extend the method?  Or do `Base.is...` (but the former style seems to be preferred here).

**Review by mikmoore — COMMENTED** [review]
Good catch on consolidating `ispos`/`isneg` from `BigInt`.
The docstrings needed an update.
Consider adding a notice to `news.md`.

File: base/int.jl [hunk 12]
```diff
@@ -139,6 +139,10 @@ iseven(n::Real) = isinteger(n) && iszero(rem(Integer(n), 2))
 signbit(x::Integer) = x < 0
 signbit(x::Unsigned) = false
 
+ispositive(x::Integer) = x > 0
+isnegative(x::Integer) = x < 0
+isnegative(x::Unsigned) = false
```
[the remark this answers is not in the document]
Comment (putianyi889) [reply base/int.jl:144]: this is aligned with `signbit` above.

File: base/number.jl [hunk 13]
```diff
@@ -136,6 +136,44 @@ true
 """
 signbit(x::Real) = x < 0
 
+"""
+    ispositive(x)
+
+Test whether `x` is positive. Falls back to [`isnan`](@ref), [`signbit`](@ref) and [`iszero`](@ref). See also [`isnegative`](@ref).
```
[the remark this answers is not in the document]
Comment (mikmoore) [reply base/number.jl]: Oops. This also needs a compatibility statement. See `?tanpi` for an example.

File: test/numbers.jl [hunk 14]
```diff
@@ -830,6 +830,63 @@ end
...
+        @test !ispositive(-1//0)
+
+        @test !isnegative(2//3)
+        @test isnegative(-2//3)
+        @test !isnegative(0//1)
+        @test !isnegative(-0//1)
+        @test !isnegative(1//0)
+        @test isnegative(-1//0)
+    end
+end
```
[the remark this answers is not in the document]
Comment (putianyi889) [reply test/numbers.jl:889]: @giordano opinions?

File: base/bool.jl [hunk 15]
```diff
@@ -154,6 +154,7 @@ abs(x::Bool) = x
 abs2(x::Bool) = x
 iszero(x::Bool) = !x
 isone(x::Bool) = x
+ispositive(x::Bool) = x
```
[the remark this answers is not in the document]
Comment (StefanKarpinski) [reply base/bool.jl:157]: I think it's simpler and clear to see that it's the identity function for booleans, but I guess the definition of `isone` is right there. You can also probably just let the compiler optimize the fallback.

[the remark this answers is not in the document]
Comment (putianyi889) [reply base/bool.jl:157]: > You can also probably just let the compiler optimize the fallback.

The compiler doesn't do that at present.
```julia
julia> code_llvm(x->x>0, (Bool,))
;  @ REPL[6]:1 within `#5`
; Function Attrs: uwtable
define i8 @"julia_#5_5646"(i8 zeroext %0) #0 {
top:
; ┌ @ operators.jl:378 within `>`
; │┌ @ promotion.jl:462 within `<`
; ││┌ @ promotion.jl:393 within `promote`
; │││┌ @ promotion.jl:370 within `_promote`
; ││││┌ @ number.jl:7 within `convert`
; │││││┌ @ boot.jl:784 within `Int64`
; ││││││┌ @ boot.jl:711 within `toInt64`
         %1 = and i8 %0, 1
; ││└└└└└
; ││ @ promotion.jl:462 within `<` @ int.jl:83
    ret i8 %1
; └└
}

julia> code_llvm(x->x, (Bool,))
;  @ REPL[7]:1 within `#7`
; Function Attrs: uwtable
define i8 @"julia_#7_5660"(i8 zeroext %0) #0 {
top:
  ret i8 %0
}
```

File: test/numbers.jl (hunk 14 above)
[the remark this answers is not in the document]
Comment (giordano) [reply test/numbers.jl:889]: Sounds good to me, although I'm not entirely sure at a quick glance why `@eval` is needed.

[the remark this answers is not in the document]
Comment (mikmoore) [reply test/numbers.jl:889]: > why `@eval` is needed.

It's just needed to make any failures clear. Compare the following two test failures, for example:
```julia-repl
julia> value = 2+0im; @test ispositive(value) === value > 0
Error During Test at REPL[108]:1
  Test threw exception
  Expression: ispositive(value) === value > 0

julia> value = 2+0im; @eval @test ispositive($value) === $value > 0
Error During Test at REPL[109]:1
  Test threw exception
  Expression: ispositive(2 + 0im) === 2 + 0im > 0
```
Notice that without using `@eval` to interpolate the value, it doesn't indicate *which* value failed.

File: base/number.jl [hunk 16]
```diff
@@ -136,6 +136,52 @@ true
 """
 signbit(x::Real) = x < 0
 
+"""
+    ispositive(x)
+
+Test whether `x > 0`. See also [`isnegative`](@ref).
+
+!!! compat "Julia 1.11"
+    This function requires at least Julia 1.11.
```
Comment (mikmoore) [inline base/number.jl:145]:
```suggestion
!!! compat "Julia 1.12"
    This function requires at least Julia 1.12.
```
The feature freeze for 1.11 was last month and it's already in alpha. I expect this to hit in 1.12.

File: base/number.jl [hunk 17]
```diff
@@ -136,6 +136,52 @@ true
...
+ispositive(x) = x > zero(x) # generic fallback
+ispositive(x::Real) = x > 0
+
+"""
+    isnegative(x)
+
+Test whether `x < 0`. See also [`ispositive`](@ref).
+
+!!! compat "Julia 1.11"
+    This function requires at least Julia 1.11.
```
Comment (mikmoore) [inline base/number.jl:168]:
```suggestion
!!! compat "Julia 1.12"
    This function requires at least Julia 1.12.
```

**Review by mikmoore — APPROVED** [review]
Pending the compat version changes I suggested, this looks fine to me. Could use a `news.md` announcement.

File: base/number.jl [hunk 18]
```diff
@@ -136,6 +136,52 @@ true
...
+
+julia> ispositive(99)
+true
+
+julia> ispositive(0.0)
+false
+```
+"""
+ispositive(x) = x > zero(x) # generic fallback
+ispositive(x::Real) = x > 0
```
Comment (giordano) [inline base/number.jl:160]: I probably missed it in the discussions above, but what would be the difference between `x > 0` and `x > zero(x)`?

File: base/bool.jl (hunk 15 above)
[the remark this answers is not in the document]
Comment (giordano) [reply base/bool.jl:157]: I think the compiler is being a bit silly here
```
julia> code_llvm(x -> x && true, (Bool,))
```
```llvm
;  @ REPL[4]:1 within `#5`
define i8 @"julia_#5_249"(i8 zeroext %0) #0 {
top:
  %1 = and i8 %0, 1
  ret i8 %1
}
```
it doesn't realise could simply return the input argument.

File: base/number.jl (hunk 18 above)
Comment (putianyi889) [reply base/number.jl:160]: It's not related to the discussions. I just realized
- similar methods only support `Real`.
- https://github.com/JuliaLang/julia/issues/35067 asked for a generic comparison againse `zero(x)` and it's also an example where `x>0` doesn't work.

Comment (mikmoore) [reply base/number.jl:160]: I guess the lowest real type on the `promote` totem pole is `Bool`, so I would suggest `x > false` as the generic case (although for most types, no promotion actually occurs). I agree that `x > zero(x)` doesn't add much and requires a type to explicitly define `zero` (which virtually all will, but it seems like an unnecessary dependency). For types requiring heap allocation (like `BigInt`, although it is covered by a separate specialization), `zero` will necessarily require an allocation whereas `0` or `false` might avoid it when mixed-type comparisons are defined.

In other words, `BigInt` is an example indicating that `x > 0`/`x > false` is likely a more performant fallback than `x > zero(x)`.

File: base/bool.jl (hunk 15 above)
[the remark this answers is not in the document]
Comment (giordano) [reply base/bool.jl:157]: Ah, that's #21712

File: base/number.jl (hunk 18 above)
Comment (giordano) [reply base/number.jl:160]: I was more wondering why having the two different methods to start with, sounds like we should a single good definition, since `Real` isn't much more specific than `Any`.

Comment (putianyi889) [reply base/number.jl:160]: as `BigInt` and `BigFloat` have specialized methods anyway, I guess using `zero(x)` universally is fine.

**Comment by nsajko** [comment]
Tests are failing:

```none
numbers                                          (16) |         failed at 2024-10-21T07:52:16.381
Test Failed at /cache/build/tester-demeter6-13/julialang/julia-master/julia-761ed61ee4/share/julia/test/numbers.jl:844
  Expression: ispositive(0) === 0 > 0
   Evaluated: false === 0 > 0
Test Failed at /cache/build/tester-demeter6-13/julialang/julia-master/julia-761ed61ee4/share/julia/test/numbers.jl:845
  Expression: ispositive(-0) === -0 > 0
   Evaluated: false === 0 > 0
[...]
```

File: base/number.jl [hunk 19]
```diff
@@ -136,6 +136,52 @@ true
...
+false
+
+julia> ispositive(99)
+true
+
+julia> ispositive(0.0)
+false
+```
+"""
+ispositive(x) = x > zero(x) # generic fallback
```
Comment (nsajko) [inline base/number.jl:159]: :-1:

Why is there an `Any` method? The definition mostly makes sense for `Real`.

Comment (putianyi889) [reply base/number.jl:159]: It is because of #35067

File: base/number.jl (hunk 18 above)
Comment (nsajko) [reply base/number.jl:160]: > some types that can behave like reals are not actually `<: Real`. `DualNumbers.Dual <: Number` being such an example.

I don't think the dual numbers are ordered though? So I guess we don't want `ispositive` or `isnegative` defined for dual numbers.

File: base/number.jl (hunk 19 above)
Comment (putianyi889) [reply base/number.jl:159]: In any case it should not belong to `number.jl`. If it has to be added, where should it go?

Comment (nsajko) [reply base/number.jl:159]: > It is because of https://github.com/JuliaLang/julia/issues/35067

I don't see the connection. That issue doesn't seem to discuss that part of the design. It barely discusses design at all.

Comment (putianyi889) [reply base/number.jl:159]: That issue discussed reasons of having the general interface.

Comment (mbauman) [reply base/number.jl:159]: I agree with @nsajko. It's always possible to add more generic functionality later, but it's impossible to go backwards.  The most generic `zero` gets is `::Number`.  If we have a function in base, then non-numeric types (like Dates) can of course choose to extend it if appropriate.

Comment (nsajko) [reply base/number.jl:159]: Oh, I get it, thanks.

**Comment by putianyi889** [comment]
doctest is failing due to #54457

**Comment by putianyi889** [comment]
one test is failing due to timeout

**Comment by putianyi889** [comment]
builds fail consistently on powerpc.

File: base/int.jl [hunk 20]
```diff
@@ -139,6 +139,10 @@ iseven(n::Real) = isinteger(n) && iszero(rem(Integer(n), 2))
 signbit(x::Integer) = x < 0
 signbit(x::Unsigned) = false
 
+ispositive(x::Integer) = x > 0
+isnegative(x::Integer) = x < 0
```
Comment (giordano) [inline base/int.jl:143]: Isn't this covered by
```julia
ispositive(x::Real) = x > 0
isnegative(x::Real) = x < 0
```
already?

**Review by giordano — COMMENTED** [review]
I think the docstrings need to be added to the manual, no?

File: test/numbers.jl [hunk 21]
```diff
@@ -830,6 +830,26 @@ end
...
+        elseif T <: Rational
+            push!(values, 1//0) # also check Infs
+        end
+        @testset "$T" begin
+            for value in values
+                @eval begin
+                    @test ispositive($value) === ($value > 0)
+                    @test ispositive(-$value) === (-$value > 0)
+                    @test isnegative($value) === ($value < 0)
+                    @test isnegative(-$value) === (-$value < 0)
```
Comment (giordano) [inline test/numbers.jl:847]: I'm missing why this needs `@eval`

File: test/numbers.jl [hunk 22]
```diff
@@ -830,6 +830,26 @@ end
     @test cmp(isless, 1, NaN) == -1
     @test cmp(isless, NaN, NaN) == 0
 end
+@testset "ispositive/isnegative" begin
+        for T in [Base.uniontypes(Base.BitInteger)..., Bool, Rational{Int}, BigInt, Base.uniontypes(Base.IEEEFloat)..., BigFloat, Missing]
```
Comment (giordano) [inline test/numbers.jl:834]: Why the 8-space indentation?   You use 4-space below.

File: test/numbers.jl [hunk 23]
```diff
@@ -830,6 +830,26 @@ end
...
+        values = T[zero(T), one(T)]
```
Comment (giordano) [inline test/numbers.jl:835]: Am I missing something or indentation here is off and the closing `end`s below are correctly indented just because above you used 8 spaces?  But probably fixing the indentation above would solve this issue too.

File: test/numbers.jl (hunk 21 above)
Comment (putianyi889) [reply test/numbers.jl:847]: https://github.com/JuliaLang/julia/pull/53677#discussion_r1533736553

Comment (giordano) [reply test/numbers.jl:847]: At least I'm consistent 🙂 Maybe this highlights the need to write down a comment to explain this, because I'm sure next time I'll re-read this code I'll ask the same question

**Comment by giordano** [comment]
> Should they be with `isless` (base.md), `isfinite` (numbers.md) or `signbit` (math.md)?

Having to choose, I'd probably pick `isless`, seems to be the somewhat closer in spirit.

**Review by giordano — APPROVED** [review]

**Review by fingolfin — APPROVED** [review]
We have had ispositive & isnegative methods in AbstractAlgebra for some time already, so this feels quite natural to me.

I've resolved a conflict in `NEWS.md` and hopefully this can be merged once CI passes?

File: base/number.jl [hunk 24]
```diff
@@ -136,6 +136,50 @@ true
...
+"""
+ispositive(x::Real) = x > 0
+
+"""
+    isnegative(x)
+
+Test whether `x < 0`. See also [`ispositive`](@ref).
+
+!!! compat "Julia 1.12"
+    This function requires at least Julia 1.12.
```
Comment (giordano) [inline base/number.jl:167]:
```suggestion
!!! compat "Julia 1.13"
    This function requires at least Julia 1.13.
```
